src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp

```diff
@@ -0,0 +1,139 @@
+    if (order == BY_SCORE) {
+        sort_results_by_score(res);
+    }
+    return res;
+}
+
+VecSimQueryResult *BF_BatchIterator::heapBasedSearch(size_t n_res) {
+    float upperBound = std::numeric_limits<float>::lowest();
+    CandidatesHeap TopCandidates;
+    // map vector's label to its index in the scores vector.
+    unordered_map<size_t, size_t> TopCandidatesIndices(n_res);
+    for (int i = 0; i < scores.size(); i++) {
+        if (isnan(scores[i].first)) {
+            continue;
+        }
+        if (TopCandidates.size() < n_res) {
+            TopCandidates.emplace(scores[i].first, scores[i].second);
+            TopCandidatesIndices[scores[i].second] = i;
+            upperBound = TopCandidates.top().first;
+        } else {
+            if (scores[i].first >= upperBound) {
+                continue;
+            } else {
+                TopCandidates.emplace(scores[i].first, scores[i].second);
+                TopCandidatesIndices[scores[i].second] = i;
+                // remove the furthest vector from the candidates and from the label->index mappings
+                TopCandidatesIndices.erase(TopCandidates.top().second);
+                TopCandidates.pop();
+                upperBound = TopCandidates.top().first;
+            }
+        }
+    }
+
+    auto *results = array_new_len<VecSimQueryResult>(TopCandidates.size(), TopCandidates.size());
+    for (int i = (int)TopCandidates.size() - 1; i >= 0; --i) {
+        VecSimQueryResult_SetId(results[i], TopCandidates.top().second);
+        VecSimQueryResult_SetScore(results[i], TopCandidates.top().first);
+        // Invalidate vector score, so we won't return it again in the next iterations.
+        scores[TopCandidatesIndices[TopCandidates.top().second]].first = INVALID_SCORE;
+        TopCandidates.pop();
+    }
+    return results;
+}
+
+VecSimQueryResult *BF_BatchIterator::selectBasedSearch(size_t n_res) {
+    // We return up to n_res vectors, the remaining vectors size is an upper bound.
+    if (n_res > scores.size()) {
+        n_res = scores.size();
+    }
+    auto n_th_element_pos = scores.begin() + (int)n_res;
+    // This will perform an in-place partition of the elements in the array, based on the
+    // n-th element as the pivot - every element with a lower will be placed before it, and
+    // all the rest will be placed after.
+    std::nth_element(scores.begin(), n_th_element_pos, scores.end(), cmp);
+
+    auto *results = array_new<VecSimQueryResult>(n_res);
+    for (size_t i = 0; i < n_res; i++) {
+        if (isnan(scores[i].first)) {
+            continue;
+        }
+        results = array_append(results, VecSimQueryResult{});
+        VecSimQueryResult_SetId(results[array_len(results) - 1], scores[i].second);
+        VecSimQueryResult_SetScore(results[array_len(results) - 1], scores[i].first);
+    }
+    // Erase the returned results slice from the scores array
+    scores.erase(scores.begin(), n_th_element_pos);
```

Comment:
didn't we say we are not erasing but moving the start of the array to `n_th_element_pos`?

Reply:
You're right. Changing this